aggregator/internal/pkg/server.go

```diff
@@ -35,6 +35,48 @@ func (agg *Aggregator) ServeOperators() error {
 	return err
 }
 
+// Waits for the arrival of task associated with signedTaskResponse and returns true on success or false on failure
+// If the task is not present in the internal map, it will try to fetch it from logs and retry.
+// The number of retries is specified by `waitForEventRetries`, and the waiting time between each by `waitForEventSleepSeconds`
+func (agg *Aggregator) waitForTaskAndFetchIfLost(signedTaskResponse *types.SignedTaskResponse) bool {
+	for i := 0; i < waitForEventRetries; i++ {
+		// Lock
+		agg.taskMutex.Lock()
+		agg.AggregatorConfig.BaseConfig.Logger.Info("- Locked Resources: Check if task is present")
+		_, ok := agg.batchesIdxByIdentifierHash[signedTaskResponse.BatchIdentifierHash]
+		// Unlock
+		agg.logger.Info("- Unlocked Resources: Check if task is present")
+		agg.taskMutex.Unlock()
+		if ok {
+			return true
+		}
+
+		// Task was not found in internal map, let's try to fetch it from logs
+		agg.logger.Info("Trying to fetch missed task from logs...")
+		batch, err := agg.avsReader.GetPendingBatchFromMerkleRoot(signedTaskResponse.BatchMerkleRoot)
+
+		if err == nil && batch != nil {
+			agg.logger.Info("Found missed task in logs with merkle root 0x%e", batch.BatchMerkleRoot)
+			// Adding new task will fail only if it already exists
+			agg.AddNewTask(batch.BatchMerkleRoot, batch.SenderAddress, batch.TaskCreatedBlock)
+			return true
+		}
+
+		if err != nil {
+			agg.logger.Warn("Error fetching task from logs: %v", err)
+		}
+
+		if batch == nil {
+			agg.logger.Info("Task not found in logs")
+		}
+
+		// Task was not found, wait and retry
+		time.Sleep(waitForEventSleepSeconds)
+	}
```

Comment:
Why a new separate retry logic here?

Reply:
It's not a new separate retry logic, it's the same one but moved to an auxiliary function. You can check [here](https://github.com/yetanotherco/aligned_layer/blob/testnet/aggregator/internal/pkg/server.go#L55) how it was before.
This retry logic is needed because operators may receive the new task event before aggregators. Very unusual, but possible.
We may refactor this in the future once [the retry refactor PR](https://github.com/yetanotherco/aligned_layer/pull/1304/files#diff-566a66135148f53094f82c777b2a3f1a3cab38cdf10feabe0790e2548df58a93R133-R142) is merged.